libs/s25main/controls/ctrlMapSelection.cpp

```diff
@@ -0,0 +1,244 @@
+    auto* res = dynamic_cast<glArchivItem_Bitmap*>(enabledMask.get());
+    RTTR_Assert(!enabledMask.get() || res);
+    mapImages.enabledMaskMemory = std::move(enabledMask);
+    return res;
+}
+
+bool ctrlMapSelection::setupMap()
+{
+    mapImages.background = loadInputImage(inputData.background);
+    mapImages.map = loadInputImage(inputData.map);
+    mapImages.missionMapMask = loadInputImage(inputData.missionMapMask);
+    mapImages.cursor = loadInputImage(inputData.cursor);
+    mapImages.conquered = loadInputImage(inputData.conquered);
+    mapImages.enabledMask = createEnabledMask(mapImages.missionMapMask->GetSize());
+
+    updateEnabledMask();
+    return mapImages.isValid();
+}
+
+bool ctrlMapSelection::IsMouseOver(const Position& mousePos) const
+{
+    return IsPointInRect(mousePos, GetDrawRect());
+}
+
+bool ctrlMapSelection::scaleX()
+{
+    auto xRatio = GetSize().x / (float)mapImages.background->GetSize().x;
+    auto yRatio = GetSize().y / (float)mapImages.background->GetSize().y;
+    return xRatio < yRatio;
+}
+
+DrawPoint ctrlMapSelection::invertScale(DrawPoint const& scaleIt)
+{
+    if(scaleX())
+        return DrawPoint(scaleIt.x * mapImages.background->GetSize().x / GetSize().x,
+                         scaleIt.y * mapImages.background->GetSize().x / GetSize().x);
+    else
+        return DrawPoint(scaleIt.x * mapImages.background->GetSize().y / GetSize().y,
+                         scaleIt.y * mapImages.background->GetSize().y / GetSize().y);
+}
+
+Extent ctrlMapSelection::scale(Extent const& scaleIt)
+{
+    if(scaleX())
+        return Extent(scaleIt.x * GetSize().x / mapImages.background->GetSize().x,
+                      scaleIt.y * GetSize().x / mapImages.background->GetSize().x);
+    else
+        return Extent(scaleIt.x * GetSize().y / mapImages.background->GetSize().y,
+                      scaleIt.y * GetSize().y / mapImages.background->GetSize().y);
+}
+
+DrawPoint ctrlMapSelection::scale(DrawPoint const& scaleIt)
+{
+    if(scaleX())
+        return DrawPoint(scaleIt.x * GetSize().x / mapImages.background->GetSize().x,
+                         scaleIt.y * GetSize().x / mapImages.background->GetSize().x);
+    else
+        return DrawPoint(scaleIt.x * GetSize().y / mapImages.background->GetSize().y,
+                         scaleIt.y * GetSize().y / mapImages.background->GetSize().y);
+}
+
+DrawPoint ctrlMapSelection::getBackgroundPosition()
+{
+    return GetDrawPos() + (GetSize() - scale(mapImages.background->GetSize())) / 2;
+}
+
+DrawPoint ctrlMapSelection::getMapOffsetRelativeToBackground()
+{
+    return scale(inputData.mapOffsetInBackground);
+}
+
+DrawPoint ctrlMapSelection::getMapPosition()
+{
+    return getBackgroundPosition() + getMapOffsetRelativeToBackground();
+}
+
+DrawPoint ctrlMapSelection::getScaledImageOriginOffset(glArchivItem_Bitmap* bitmap)
+{
+    return DrawPoint(bitmap->getNx(), bitmap->getNy()) - scale(DrawPoint(bitmap->getNx(), bitmap->getNy()));
```

Comment:
```suggestion
    return bitmap->GetOrigin() - scale(bitmap->GetOrigin());
```

And maybe add a comment why you do this, although I understand it when looking at the usage

Maybe it is even better to encapsulate the whole draw call here (calling it `drawImageOnMap` or so):

Those 2 uses look similar enough:
```
            mapImages.conquered->DrawFull(Rect(getMapPosition() + 
scale(inputData.missionDescriptionList[idx].ankerPos) +
getScaledImageOriginOffset(mapImages.conquered),
scale(mapImages.conquered->GetSize())));

            mapImages.cursor->DrawFull(Rect(getMapPosition() +
scale(currentSelectionPos) +
getScaledImageOriginOffset(mapImages.cursor),
scale(mapImages.cursor->GetSize())));
```